src/storage/UpdateVertexProcessor.cpp

```diff
@@ -0,0 +1,366 @@
+
+    if (iter && iter->valid()) {
+        auto reader = RowReader::getTagPropReader(this->schemaMan_,
+                                                  iter->val(),
+                                                  this->spaceId_,
+                                                  tagId);
+        const auto constSchema = reader->getSchema();
+        for (auto& prop : props) {
+            auto res = RowReader::getPropByName(reader.get(), prop.prop_.name);
+            if (!ok(res)) {
+                VLOG(1) << "Skip the bad value for tag: " << tagId
+                        << ", prop " << prop.prop_.name;
+                return kvstore::ResultCode::ERR_UNKNOWN;
+            }
+            auto&& v = value(std::move(res));
+            tagFilters_.emplace(std::make_pair(tagId, prop.prop_.name), v);
+        }
+        if (updateTagIds_.find(tagId) != updateTagIds_.end()) {
+            auto schema = std::const_pointer_cast<meta::SchemaProviderIf>(constSchema);
+            auto updater =
+                std::unique_ptr<RowUpdater>(new RowUpdater(std::move(reader), schema));
+            pendingUpdater_[tagId] = std::make_unique<KeyUpdaterPair>();
+            auto it = pendingUpdater_.find(tagId);
+            it->second->key = iter->key().toString();
+            it->second->updater = std::move(updater);
+        }
+    } else if (insertable_ && updateTagIds_.find(tagId) != updateTagIds_.end()) {
+        resp_.set_upsert(true);
+        const auto constSchema = this->schemaMan_->getTagSchema(this->spaceId_, tagId);
+        for (auto index = 0UL; index < constSchema->getNumFields(); index++) {
+            auto propName = std::string(constSchema->getFieldName(index));
+            VariantType v;
+            switch (constSchema->getFieldType(index).get_type()) {
+                case nebula::cpp2::SupportedType::INT:
+                case nebula::cpp2::SupportedType::VID:
+                case nebula::cpp2::SupportedType::TIMESTAMP:
+                    v = folly::to<int64_t>(0);
+                    break;
+                case nebula::cpp2::SupportedType::STRING:
+                    v = folly::to<std::string>("");
+                    break;
+                case nebula::cpp2::SupportedType::FLOAT:
+                    v = folly::to<float>(0.0f);
+                    break;
+                case nebula::cpp2::SupportedType::DOUBLE:
+                    v = folly::to<double>(0.0f);
+                    break;
+                case nebula::cpp2::SupportedType::BOOL:
+                    v = folly::to<bool>(false);
+                    break;
+                default:
+                    LOG(FATAL) << "Unimplemented";
+                    return kvstore::ResultCode::ERR_UNKNOWN;
+            }
+            tagFilters_.emplace(std::make_pair(tagId, propName), v);
+        }
+        auto schema = std::const_pointer_cast<meta::SchemaProviderIf>(constSchema);
+        auto updater = std::unique_ptr<RowUpdater>(new RowUpdater(schema));
```

Comment:
Why not use a RowWriter here? Is this because keep a same interface?

Reply:
Because there are maybe some fields to be updated in the `updateAndWriteBack()` function.  RowUpdate also can to be encoded to string.